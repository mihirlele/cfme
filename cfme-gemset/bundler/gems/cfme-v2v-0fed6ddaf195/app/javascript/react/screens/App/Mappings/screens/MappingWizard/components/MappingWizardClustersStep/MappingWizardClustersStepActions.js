import URI from 'urijs';
import API from '../../../../../../../../common/API';
import { FETCH_V2V_SOURCE_CLUSTERS, QUERY_V2V_PROVIDERS } from './MappingWizardClustersStepConstants';

export { showAlertAction, hideAlertAction } from '../../MappingWizardActions';

const _getSourceClustersActionCreator = url => dispatch =>
  dispatch({
    type: FETCH_V2V_SOURCE_CLUSTERS,
    payload: API.get(url)
  });

export const fetchSourceClustersAction = url => {
  const uri = new URI(url);
  return _getSourceClustersActionCreator(uri.toString());
};

// ****************************************************************************
// QUERY_V2V_PROVIDERS
// ****************************************************************************
const _queryProvidersActionCreator = (url, providerIds) => dispatch =>
  dispatch({
    type: QUERY_V2V_PROVIDERS,
    payload: API.post(url, {
      action: 'query',
      resources: providerIds.map(id => ({ id }))
    })
  });

export const queryProvidersAction = (url, providerIds) => {
  const uri = new URI(url);
  uri.addSearch({ attributes: 'authentications' });

  return _queryProvidersActionCreator(uri.toString(), providerIds);
};
