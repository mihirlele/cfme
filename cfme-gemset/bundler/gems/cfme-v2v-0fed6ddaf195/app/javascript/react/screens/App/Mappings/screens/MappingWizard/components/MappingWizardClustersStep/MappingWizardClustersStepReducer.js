import Immutable from 'seamless-immutable';

import { FETCH_V2V_SOURCE_CLUSTERS, QUERY_V2V_PROVIDERS } from './MappingWizardClustersStepConstants';
import { V2V_MAPPING_WIZARD_EXITED } from '../../../../screens/MappingWizard/MappingWizardConstants';

export const initialState = Immutable({
  sourceClusters: [],
  isFetchingSourceClusters: false,
  isRejectedSourceClusters: false,
  errorSourceClusters: null,
  targetClusters: [],
  isFetchingTargetClusters: false,
  isRejectedTargetClusters: false,
  errorTargetClusters: null,
  providers: [],
  isQueryingProviders: false,
  isRejectedQueryProviders: false,
  errorQueryProviders: null
});

export default (state = initialState, action) => {
  switch (action.type) {
    case `${FETCH_V2V_SOURCE_CLUSTERS}_PENDING`:
      return state.set('isFetchingSourceClusters', true).set('isRejectedSourceClusters', false);
    case `${FETCH_V2V_SOURCE_CLUSTERS}_FULFILLED`:
      if (action.payload.data && action.payload.data.resources && action.payload.data.resources[0]) {
        return state
          .set('sourceClusters', action.payload.data.resources)
          .set('isRejectedSourceClusters', false)
          .set('isFetchingSourceClusters', false);
      }
      return state
        .set('sourceClusters', [])
        .set('isFetchingSourceClusters', false)
        .set('isRejectedSourceClusters', false);
    case `${FETCH_V2V_SOURCE_CLUSTERS}_REJECTED`:
      return state
        .set('errorSourceClusters', action.payload)
        .set('isRejectedSourceClusters', true)
        .set('isFetchingSourceClusters', false);

    case `${QUERY_V2V_PROVIDERS}_PENDING`:
      return state.set('isQueryingProviders', true).set('isRejectedQueryProviders', false);
    case `${QUERY_V2V_PROVIDERS}_FULFILLED`:
      return state
        .set('providers', action.payload.data.results)
        .set('isQueryingProviders', false)
        .set('isRejectedQueryProviders', false)
        .set('errorQueryProviders', null);
    case `${QUERY_V2V_PROVIDERS}_REJECTED`:
      return state
        .set('errorQueryProviders', action.payload)
        .set('isRejectedQueryProviders', true)
        .set('isQueryingProviders', false);
    case V2V_MAPPING_WIZARD_EXITED:
      return initialState;
    default:
      return state;
  }
};
