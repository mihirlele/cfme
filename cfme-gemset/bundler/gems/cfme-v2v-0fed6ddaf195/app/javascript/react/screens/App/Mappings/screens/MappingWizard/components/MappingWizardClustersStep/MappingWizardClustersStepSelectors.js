import { CONVERSION_HOST_TYPES } from '../../../../../../../../common/constants';

export const sourceClustersFilter = (sourceClustersToFilter, clustersStepMappings) => {
  const mappedSourceClusters = clustersStepMappings.reduce(
    (sourceClusters, targetClusterWithSourceClusters) => sourceClusters.concat(targetClusterWithSourceClusters.nodes),
    []
  );

  return sourceClustersToFilter.filter(
    sourceCluster => !mappedSourceClusters.some(mappedSourceCluster => mappedSourceCluster.id === sourceCluster.id)
  );
};

export const conversionHostsFilter = (conversionHosts, providerType) =>
  conversionHosts.filter(host => host.resource && host.resource.type === CONVERSION_HOST_TYPES[providerType]);
