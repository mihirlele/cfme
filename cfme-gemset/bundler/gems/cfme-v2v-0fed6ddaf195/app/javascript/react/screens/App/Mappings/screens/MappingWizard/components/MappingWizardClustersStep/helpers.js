export const getProviderIds = clusters => {
  const providerIds = new Set();
  clusters.forEach(cluster => providerIds.add(cluster.ems_id));
  return [...providerIds];
};
