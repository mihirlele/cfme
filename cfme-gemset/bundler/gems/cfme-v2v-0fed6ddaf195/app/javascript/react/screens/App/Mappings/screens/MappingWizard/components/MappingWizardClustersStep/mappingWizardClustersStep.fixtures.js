import Immutable from 'seamless-immutable';

export const sourceClusters = Immutable({
  name: 'clusters',
  count: 5,
  subcount: 1,
  subquery_count: 1,
  pages: 1,
  resources: [
    {
      href: 'http://localhost:3000/api/clusters/10000000000001',
      id: '1',
      name: 'VMWare Cluster1',
      ems_id: '10000000000012',
      created_on: '2016-07-15T18:28:16Z',
      updated_on: '2016-07-15T18:28:16Z',
      uid_ems: '00000002-0002-0002-0002-00000000024b',
      ha_enabled: null,
      ha_admit_control: null,
      ha_max_failures: null,
      drs_enabled: null,
      drs_automation_level: null,
      drs_migration_threshold: null,
      last_perf_capture_on: null,
      ems_ref_obj: '/api/clusters/00000002-0002-0002-0002-00000000024b',
      effective_cpu: null,
      effective_memory: null,
      ems_ref: '/api/clusters/00000002-0002-0002-0002-00000000024b',
      type: null,
      ext_management_system: { emstype: 'vmwarews' },
      v_parent_datacenter: 'Datacenter'
    },
    {
      href: 'http://localhost:3000/api/clusters/10000000000002',
      id: '2',
      name: 'VMWare Cluster2',
      ems_id: '10000000000012',
      created_on: '2016-07-15T18:28:16Z',
      updated_on: '2016-07-15T18:28:16Z',
      uid_ems: '00000002-0002-0002-0002-00000000024b',
      ha_enabled: null,
      ha_admit_control: null,
      ha_max_failures: null,
      drs_enabled: null,
      drs_automation_level: null,
      drs_migration_threshold: null,
      last_perf_capture_on: null,
      ems_ref_obj: '/api/clusters/00000002-0002-0002-0002-00000000024b',
      effective_cpu: null,
      effective_memory: null,
      ems_ref: '/api/clusters/00000002-0002-0002-0002-00000000024b',
      type: null,
      ext_management_system: { emstype: 'vmwarews' },
      v_parent_datacenter: 'Datacenter'
    },
    {
      href: 'http://localhost:3000/api/clusters/10000000000003',
      id: '3',
      name: 'VMWare Cluster3',
      ems_id: '10000000000012',
      created_on: '2016-07-15T18:28:16Z',
      updated_on: '2016-07-15T18:28:16Z',
      uid_ems: '00000002-0002-0002-0002-00000000024b',
      ha_enabled: null,
      ha_admit_control: null,
      ha_max_failures: null,
      drs_enabled: null,
      drs_automation_level: null,
      drs_migration_threshold: null,
      last_perf_capture_on: null,
      ems_ref_obj: '/api/clusters/00000002-0002-0002-0002-00000000024b',
      effective_cpu: null,
      effective_memory: null,
      ems_ref: '/api/clusters/00000002-0002-0002-0002-00000000024b',
      type: null,
      ext_management_system: { emstype: 'vmwarews' },
      v_parent_datacenter: 'Datacenter'
    }
  ]
});

export const requestSourceClustersData = {
  method: 'GET',
  fetchSourceClustersUrl: '/api/dummyProviders',
  response: { data: sourceClusters }
};

export const initialState = Immutable({
  sourceClusters: [],
  fetchSourceClustersUrl: requestSourceClustersData.fetchSourceClustersUrl
});
