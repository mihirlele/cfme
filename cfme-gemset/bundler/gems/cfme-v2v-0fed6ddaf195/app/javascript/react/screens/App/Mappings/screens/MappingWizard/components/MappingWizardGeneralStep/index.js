import { connect } from 'react-redux';

import MappingWizardGeneralStep from './MappingWizardGeneralStep';
import * as MappingWizardGeneralStepActions from './MappingWizardGeneralStepActions';
import { showAlertAction, hideAlertAction } from '../../MappingWizardActions';

import reducer from './MappingWizardGeneralStepReducer';

export const reducers = { mappingWizardGeneralStep: reducer };

const mapStateToProps = ({
  mappingWizardGeneralStep: { editingMapping, conversionHosts },
  overview: { transformationMappings },
  form: { mappingWizardGeneralStep }
}) => ({
  editingMapping,
  conversionHosts,
  transformationMappings,
  targetProvider:
    mappingWizardGeneralStep && mappingWizardGeneralStep.values && mappingWizardGeneralStep.values.targetProvider
});

const actions = { showAlertAction, hideAlertAction, ...MappingWizardGeneralStepActions };

export default connect(
  mapStateToProps,
  actions
)(MappingWizardGeneralStep);
