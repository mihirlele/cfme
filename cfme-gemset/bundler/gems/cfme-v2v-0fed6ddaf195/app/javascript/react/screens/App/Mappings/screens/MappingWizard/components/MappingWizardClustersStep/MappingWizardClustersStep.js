import React from 'react';
import PropTypes from 'prop-types';
import { Button, noop } from 'patternfly-react';
import { Field, reduxForm } from 'redux-form';
import { length } from 'redux-form-validators';

import ClustersStepForm from './components/ClustersStepForm/ClustersStepForm';
import { QUERY_PROVIDERS_URL } from './MappingWizardClustersStepConstants';
import { FETCH_TARGET_COMPUTE_URLS } from '../../../../../../../../redux/common/targetResources/targetResourcesConstants';
import { getProviderIds } from './helpers';
import { OPENSTACK } from '../../MappingWizardConstants';

class MappingWizardClustersStep extends React.Component {
  componentDidMount() {
    const { targetProvider, ospConversionHosts, showAlertAction } = this.props;
    this.fetchClusters();
    if (targetProvider === OPENSTACK && ospConversionHosts.length === 0) {
      showAlertAction(
        __('No OpenStack conversion hosts are configured. You can continue to create an infrastructure mapping, but you must configure a conversion host before migration execution.'), // prettier-ignore
        'warning'
      );
    }
  }

  componentWillUnmount() {
    this.props.hideAlertAction();
  }

  fetchClusters = () => {
    const {
      fetchSourceClustersUrl,
      fetchSourceClustersAction,
      fetchTargetComputeUrls,
      fetchTargetClustersAction,
      targetProvider,
      queryProvidersAction,
      queryProvidersUrl
    } = this.props;

    fetchSourceClustersAction(fetchSourceClustersUrl);
    fetchTargetClustersAction(fetchTargetComputeUrls[targetProvider]).then(result => {
      const safeToProceed = result.value && result.value.data && result.value.data.resources.length > 0;
      if (targetProvider === OPENSTACK && safeToProceed) {
        queryProvidersAction(queryProvidersUrl, getProviderIds(result.value.data.resources));
      }
    });
  };

  render() {
    const {
      isFetchingSourceClusters,
      sourceClusters,
      isFetchingTargetClusters,
      targetClusters,
      isRejectedSourceClusters,
      isRejectedTargetClusters,
      targetProvider,
      rhvConversionHosts,
      providers,
      isQueryingProviders
    } = this.props;

    if (isRejectedSourceClusters || isRejectedTargetClusters) {
      return (
        <div className="wizard-pf-complete blank-slate-pf">
          <div className="wizard-pf-success-icon">
            <span className="pficon pficon-error-circle-o" />
          </div>
          <h3 className="blank-slate-pf-main-action">{__('Error Retrieving Clusters')}</h3>
          <p className="blank-slate-pf-secondary-action">
            {__("We're sorry, something went wrong. Please try again.")}
          </p>
          <Button bsStyle="primary" onClick={this.fetchClusters}>
            {__('Retry')}
          </Button>
        </div>
      );
    }
    return (
      <Field
        name="clusterMappings"
        component={ClustersStepForm}
        sourceClusters={sourceClusters}
        targetClusters={targetClusters}
        validate={[length({ min: 1 })]}
        isFetchingSourceClusters={isFetchingSourceClusters}
        isFetchingTargetClusters={isFetchingTargetClusters}
        targetProvider={targetProvider}
        rhvConversionHosts={rhvConversionHosts}
        providers={providers}
        isQueryingProviders={isQueryingProviders}
      />
    );
  }
}

MappingWizardClustersStep.propTypes = {
  fetchSourceClustersUrl: PropTypes.string,
  fetchSourceClustersAction: PropTypes.func,
  fetchTargetComputeUrls: PropTypes.object,
  fetchTargetClustersAction: PropTypes.func,
  showAlertAction: PropTypes.func,
  hideAlertAction: PropTypes.func,
  sourceClusters: PropTypes.arrayOf(PropTypes.object),
  targetClusters: PropTypes.arrayOf(PropTypes.object),
  isFetchingSourceClusters: PropTypes.bool,
  isFetchingTargetClusters: PropTypes.bool,
  isRejectedSourceClusters: PropTypes.bool,
  isRejectedTargetClusters: PropTypes.bool,
  targetProvider: PropTypes.string,
  rhvConversionHosts: PropTypes.array,
  ospConversionHosts: PropTypes.array,
  providers: PropTypes.array,
  isQueryingProviders: PropTypes.bool,
  queryProvidersAction: PropTypes.func,
  queryProvidersUrl: PropTypes.string
};
MappingWizardClustersStep.defaultProps = {
  fetchSourceClustersAction: noop,
  fetchTargetClustersAction: noop,
  isFetchingSourceClusters: true,
  isFetchingTargetClusters: true,
  isRejectedSourceClusters: false,
  isRejectedTargetClusters: false,
  targetProvider: '',
  rhvConversionHosts: [],
  ospConversionHosts: [],
  fetchSourceClustersUrl:
    '/api/clusters?expand=resources' +
    '&attributes=ext_management_system.emstype,v_parent_datacenter,ext_management_system.name' +
    '&filter[]=ext_management_system.emstype=vmwarews',
  fetchTargetComputeUrls: FETCH_TARGET_COMPUTE_URLS,
  queryProvidersUrl: QUERY_PROVIDERS_URL
};

export default reduxForm({
  form: 'mappingWizardClustersStep',
  destroyOnUnmount: false
})(MappingWizardClustersStep);
