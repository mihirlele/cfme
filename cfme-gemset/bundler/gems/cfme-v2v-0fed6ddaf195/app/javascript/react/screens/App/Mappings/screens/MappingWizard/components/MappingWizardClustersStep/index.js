import { connect } from 'react-redux';

import MappingWizardClustersStep from './MappingWizardClustersStep';
import * as MappingWizardClustersStepActions from './MappingWizardClustersStepActions';
import { fetchTargetClustersAction } from '../../../../../../../../redux/common/targetResources/targetResourcesActions';
import { createClusterMappings } from './components/ClustersStepForm/helpers';
import { getTransformationMappingItemsBySourceType } from '../../helpers';
import { TRANSFORMATION_MAPPING_ITEM_SOURCE_TYPES, OPENSTACK, RHV } from '../../MappingWizardConstants';
import { conversionHostsFilter } from './MappingWizardClustersStepSelectors';

import reducer from './MappingWizardClustersStepReducer';

export const reducers = { mappingWizardClustersStep: reducer };

const mapStateToProps = (
  {
    mappingWizardClustersStep,
    mappingWizardGeneralStep: { editingMapping, conversionHosts },
    form,
    targetResources: { isFetchingTargetClusters, isRejectedTargetClusters, targetClusters }
  },
  ownProps
) => ({
  ...mappingWizardClustersStep,
  ...ownProps.data,
  isFetchingTargetClusters,
  isRejectedTargetClusters,
  targetClusters,
  targetProvider: form.mappingWizardGeneralStep.values.targetProvider,
  rhvConversionHosts: conversionHostsFilter(conversionHosts, RHV),
  ospConversionHosts: conversionHostsFilter(conversionHosts, OPENSTACK),
  initialValues: {
    clusterMappings: editingMapping
      ? createClusterMappings(
          getTransformationMappingItemsBySourceType(TRANSFORMATION_MAPPING_ITEM_SOURCE_TYPES.cluster, editingMapping),
          targetClusters,
          mappingWizardClustersStep.sourceClusters
        )
      : []
  },
  enableReinitialize: true,
  keepDirtyOnReinitialize: true
});

const mergeProps = (stateProps, dispatchProps, ownProps) => Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  { ...MappingWizardClustersStepActions, fetchTargetClustersAction },
  mergeProps
)(MappingWizardClustersStep);
