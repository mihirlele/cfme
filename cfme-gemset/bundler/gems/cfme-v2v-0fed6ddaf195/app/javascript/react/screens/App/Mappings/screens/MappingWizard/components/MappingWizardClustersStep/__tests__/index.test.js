import React from 'react';
import { reducer as formReducer } from 'redux-form';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import { generateStore } from '../../../../../../common/testReduxHelpers';
import { initialState } from '../mappingWizardClustersStep.fixtures';
import { initialState as generalStepInitialState } from '../../MappingWizardGeneralStep/MappingWizardGeneralStepReducer';
import targetResourcesReducer, {
  initialState as targetResourcesInitialState
} from '../../../../../../../../../redux/common/targetResources/targetResourcesReducer';
import { reducers as generalStepReducers } from '../../MappingWizardGeneralStep/index';
import MappingWizardClustersStep from '../MappingWizardClustersStep';
import MappingWizardClustersStepContainer, { reducers } from '../index';

describe('Mapping Wizard integration test', () => {
  const store = generateStore(
    { ...reducers, ...generalStepReducers, form: formReducer, targetResources: targetResourcesReducer },
    {
      mappingWizardClustersStep: initialState,
      mappingWizardGeneralStep: generalStepInitialState,
      form: { mappingWizardGeneralStep: { values: { targetProvider: 'rhevm' } } },
      targetResources: targetResourcesInitialState
    }
  );

  const mountComponent = () =>
    mount(
      <Provider store={store}>
        <MappingWizardClustersStepContainer />
      </Provider>
    );

  it('should mount the MappingWizardClusterStep with mapStateToProps reduced', () => {
    const wrapper = mountComponent();

    // query the unconnected component to assert reduced props
    const component = wrapper.find(MappingWizardClustersStep);
    expect(component.props()).toMatchSnapshot();

    // isFetching set true after fetch called in `componentDidMount`
    expect(component.props().isFetchingSourceClusters).toBeTruthy();
    expect(component.props().isFetchingTargetClusters).toBeTruthy();
  });
});
