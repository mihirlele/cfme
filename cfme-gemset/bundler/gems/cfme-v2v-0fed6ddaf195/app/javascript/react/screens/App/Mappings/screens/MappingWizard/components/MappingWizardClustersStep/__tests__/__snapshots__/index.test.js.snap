// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mapping Wizard integration test should mount the MappingWizardClusterStep with mapStateToProps reduced 1`] = `
Object {
  "enableReinitialize": true,
  "fetchSourceClustersAction": [Function],
  "fetchSourceClustersUrl": "/api/dummyProviders",
  "fetchTargetClustersAction": [Function],
  "hideAlertAction": [Function],
  "initialValues": Object {
    "clusterMappings": Array [],
  },
  "isFetchingSourceClusters": true,
  "isFetchingTargetClusters": true,
  "isRejectedSourceClusters": false,
  "isRejectedTargetClusters": false,
  "keepDirtyOnReinitialize": true,
  "ospConversionHosts": Array [],
  "queryProvidersAction": [Function],
  "rhvConversionHosts": Array [],
  "showAlertAction": [Function],
  "sourceClusters": Array [],
  "targetClusters": Array [],
  "targetProvider": "rhevm",
}
`;
